src/intervals/gff3/gff3.jl

```diff
@@ -0,0 +1,158 @@
+
+# Generic Feature Format Version 3 (GFF3)
+# =======================================
+#
+# https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md
+
+using URIParser: unescape
+
+"""
+Map strings to one or more strings avoiding allocation and deallocation as much
+as possible.
+"""
+type GFFAttributes <:
+        Associative{StringField, Union{StringField, Vector{StringField}}}
+    # map key to corresponding index in data and used
+    indexes::Dict{StringField, Int}
+
+    data::Vector{Vector{StringField}}
+
+    # how many fields in data[i] are used
+    used::Vector{Int}
+
+    function GFFAttributes()
+        return new(Dict{StringField, Int}(), Vector{StringField}[], Int[])
+    end
+end
+
+function Base.length(attrs::GFFAttributes)
+    return sum(attrs.used)
+end
+
+function Base.empty!(attrs::GFFAttributes)
+    fill!(attrs.used, 0)
+end
+
+function Base.haskey(attrs::GFFAttributes, key::StringField)
+    i = get(attrs.indexes, key, 0)
+    return i != 0 && attrs.used[i] > 0
+end
+
+function Base.haskey(attrs::GFFAttributes, key_::String)
+    key = StringField(key_.data, 1:length(key_.data))
+    return haskey(attrs, key)
+end
+
+function Base.getindex(attrs::GFFAttributes, key::StringField)
+    i = get(attrs.indexes, key, 0)
+    if i == 0 || attrs.used[i] == 0
+        throw(KeyError(key))
+    end
+    if length(attrs.data[i]) == 1
+        return attrs.data[i][1]
+    else
+        return attrs.data[i]
+    end
+end
+
+function Base.getindex(attrs::GFFAttributes, key_::String)
+    key = StringField(key_.data, 1:length(key_.data))
+    return getindex(attrs, key)
+end
+
+function pushindex!(attrs::GFFAttributes, key::StringField,
```

Comment:
I'm not sure what to call this function. `GFFAttributes` is basically a multi-map, so `pushindex!` was meant to imply something between `setindex!` and `push!`.